ukify: fixes with kernel compression

bluca opened the pull request.

DaanDeMeyer commented on src/ukify/ukify.py:
```diff
@@ -1297,7 +1297,18 @@ def make_uki(opts: UkifyConfig) -> None:
+            pefile.PE(linux, fast_load=True)
+        except pefile.PEFormatError:
+            decompressed = maybe_decompress(linux)
+            if decompressed:
+                linux = Path(tempfile.NamedTemporaryFile(prefix='linux-decompressed').name)
+                linux.write_bytes(decompressed)
```
Might as well check if it's a PE file again now?

bluca replied:
aren't elf files also supported?

yuwata reviewed (commented):
LGTM.
Is it possible to add test cases for the issues, especially for compression and decompression?

keszybz reviewed (approved):
LGTM.

keszybz commented:
Hmm, we now potentially decompress the file twice: once to scrape the version and once to put in UKI. With a typical kernel, this takes about a fraction of a second. But maybe with a bigger kernel built with `allyesconfig` this could be a few times more. So in the cases where we're going to decompress the file before putting it in the UKI, I think  we should write the result to a tempfile on the first decompress and not do the decompression twice.

Also, can you please add a log line that says that we're writing the decompressed kernel to the UKI? This  might not be obvious, since there's many different types of "compression" and some we leave be, and the user might be surprised if the output is significantly larger than the inputs.

bluca commented on src/ukify/ukify.py:
```diff
@@ -195,8 +195,8 @@ def get_zboot_kernel(f: IO[bytes]) -> bytes:
     elif comp_type.startswith(b'zstd22'):
-        zstd = try_import('zstd')
-        return cast(bytes, zstd.uncompress(f.read(size)))
+        zstd = try_import('zstandard')
```
Sorry no idea, I just found this by searching and came across the issue mentioned in the commit message

bluca commented:
> Hmm, we now potentially decompress the file twice: once to scrape the version and once to put in UKI. With a typical kernel, this takes about a fraction of a second. But maybe with a bigger kernel built with `allyesconfig` this could be a few times more. So in the cases where we're going to decompress the file before putting it in the UKI, I think we should write the result to a tempfile on the first decompress and not do the decompression twice.

It is already only decompressed once - if you check the commit, it changes the `linux` variable and then I also changed the bit that does uname parsing later to use it. So the second time it will see that it's a kernel and won't do any decompression - it can't anyway, as it's no longer operating on a compressed file.

> Also, can you please add a log line that says that we're writing the decompressed kernel to the UKI? This might not be obvious, since there's many different types of "compression" and some we leave be, and the user might be surprised if the output is significantly larger than the inputs.

Added

bluca commented:
> Is it possible to add test cases for the issues, especially for compression and decompression?

To be meaningful you'd need an actual arm kernel with the bad build options, and I honestly can't be bothered to set that up. The decompression was already in a bad state for zstd anyway, so it's not making it any worse

keszybz commented:
LGTM.
